### PR TITLE
Convert AOT to TECompiler

### PR DESCRIPTION
**Pull request by Mousius**
This removes the dependency on "compile_engine.h" from aot_executor_codegen.cc. This required a few changes to how AOT was operating:
* AOT run_model is now based on the post lowering main_module
* AOTOnDemandAllocator is ran twice to ensure SIDs are updated post-lowering
* Moved to using tec::UpdateFunctionMetadata

Tests are passing, but would appreciate other validation :smile_cat:

### CODE REVIEW

**Comment by Mousius** [comment]
CC: @areusch @mbs-octoml

File: src/relay/backend/aot_executor_codegen.cc
Comment (mbs-octoml) [inline src/relay/backend/aot_executor_codegen.cc:651]: Feel free to leave a TODO(mbs) to remove this reconstruction since I'm trying to replace these Expr->Storage and Expr->Device side maps with attrs.

**Review by mbs-octoml — APPROVED** [review]
FWIW LGTM  though I'm still spinning up on the memory planning machinery.

File: src/relay/backend/aot_executor_codegen.cc
Comment (Mousius) [reply src/relay/backend/aot_executor_codegen.cc:644]: I think it's ok without an explicit TODO here as it makes sense to replan the allocations? I appreciate it'll get improved later :smile_cat:

Comment (mbs-octoml) [reply src/relay/backend/aot_executor_codegen.cc:644]: Sure, but then can you comment the storage map should be morally the same as the original, just with the keys updated to follow along with the rewritten primitive calls. Or at least that's what I think should be happening, is that right?

Comment (Mousius) [reply src/relay/backend/aot_executor_codegen.cc:645]: Yip, that's correct, I've updated the comment to clarify that - what do you think now?

File: src/relay/backend/aot_executor_codegen.cc
Comment (jroesch) [inline src/relay/backend/aot_executor_codegen.cc:607]: @mbs-octoml I don't think @Mousius needs to do this in this patch, but this is where we should split device planning and storage planning. I think we can remove the need to pre-storage plan at all if we can obtain the device information pre-lowering, then storage plan after the lowering.

File: src/relay/backend/aot_executor_codegen.cc
Comment (jroesch) [inline src/relay/backend/aot_executor_codegen.cc:700]: Is it possible for us to remove the specialized `main` functions `FunctionInfo` field in LoweredModule? They should all be uniformly set inside of the lowering, and I think the specialized `main` is a hold over from Lily's code that could be cleaned up.

**Comment by jroesch** [comment]
Overall LGTM to me, once you have resolved everyone's comments I can do another pass and give a thumbs up for merge.

**Comment by jroesch** [comment]
Sorry about that, hit the wrong button 😭 reopened the PR.

File: src/relay/backend/aot_executor_codegen.cc
Comment (Mousius) [reply src/relay/backend/aot_executor_codegen.cc:665]: That field is still useful, and we can unify it between Graph and AOT to refactor altogether later; I've just pushed up a change to use the `main_func_info` attribute on the `LoweredModule` rather than the AOT bespoke way of processing this information so it can be refactored in unison later.

**Comment by Mousius** [comment]
@jroesch / @mbs-octoml could you do another pass and give approvals or change suggestions? :smile_cat:

File: src/relay/backend/aot_executor_codegen.cc
Comment (jroesch) [inline src/relay/backend/aot_executor_codegen.cc:604]: @mbs-octoml I think my point is we should be able to do like `lowered_module.main_module["main"].GetAttr("function_info")`. Maybe your point is to clean this up in future patch after landing this?

**Review by jroesch — APPROVED** [review]
LGTM, thanks for doing this, we are getting closer to being able to unify all the flows.

File: src/relay/backend/aot_executor_codegen.cc
Comment (mbs-octoml) [reply src/relay/backend/aot_executor_codegen.cc:612]: lgtm

File: src/relay/backend/aot_executor_codegen.cc
Comment (mbs-octoml) [reply src/relay/backend/aot_executor_codegen.cc:573]: Got it.

File: src/relay/backend/aot_executor_codegen.cc
Comment (manupa-arm) [inline src/relay/backend/aot_executor_codegen.cc:463]: I think we should maintain this check -- maybe inside CreateFuncCall ?

**Review by manupa-arm — COMMENTED** [review]
Just one comment! rest LGTM.

File: src/relay/backend/aot_executor_codegen.cc
Comment (Mousius) [reply src/relay/backend/aot_executor_codegen.cc:463]: Post-lowering every function is essentially a `GlobalVar` so this path was never called. If there's a test case that shows this I can re-introduce it.

Comment (manupa-arm) [reply src/relay/backend/aot_executor_codegen.cc:463]: I think this is an assumption turned assertion that all calls are made primitive. This guarantees the relay lowering is done, before the respective executor codegen is invoked.

But I see your point -- its was never checked as it should've been. Maybe its worth checking the function attached to GlobalVar has this property?

Comment (Mousius) [reply src/relay/backend/aot_executor_codegen.cc:463]: Everything is already lowered at this point as it's been through `LowerTE` before this runs, so we don't have to make the assumption - it's guaranteed :smile_cat: 

I'd also suggest that we don't add defensive code which we can't craft a way to invoke?

Comment (manupa-arm) [reply src/relay/backend/aot_executor_codegen.cc:463]: Ah I see.

So its kind of passed onto LowerTE.

Yeah, if the check itself is not invoked then no point of having it there.

LGTM.

**Review by manupa-arm — APPROVED** [review]
LGTM

**Pull request merged**
